### PR TITLE
WIP: HTTPS API endpoint support

### PR DESCRIPTION
**Pull request by dgrisham**
This PR addresses the issue https://github.com/ipfs/ipfs-cluster/issues/120. The current implementation builds, but has not yet been tested. The default TLS configuration used was copied from https://github.com/denji/golang-tls#perfect-ssl-labs-score-with-go and has not yet been verified outside of that source.

### CODE REVIEW

File: pnet_test.go [hunk 1]
```diff
@@ -3,31 +3,31 @@ package ipfscluster
 import "testing"
 
 func TestClusterSecretFormat(t *testing.T) {
-    goodSecret := "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
-    emptySecret := ""
-    tooShort := "0123456789abcdef"
-    tooLong := "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0"
-    unsupportedChars := "0123456789abcdef0123456789!!!!!!0123456789abcdef0123456789abcdef"
+	goodSecret := "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
```
Comment (hsanjuan) [inline pnet_test.go]: Hm if we are to be scrict, this belongs on a different PR...

File: api/restapi/restapi.go [hunk 2]
```diff
@@ -66,14 +67,39 @@ type peerAddBody struct {
...
+		tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+	},
+}
+
 // NewRESTAPI creates a new REST API component. It receives
-// the multiaddress on which the API listens.
-func NewRESTAPI(apiMAddr ma.Multiaddr) (*RESTAPI, error) {
+// the multiaddress on which the API listens, as well as (potentially empty)
+// paths to certificate and private key files for TLS usage
+func NewRESTAPI(apiMAddr ma.Multiaddr, certFile, keyFile string) (*RESTAPI, error) {
```
Comment (hsanjuan) [inline api/restapi/restapi.go]: It's not a good practice to change the API (I mean, as in the signature of a public function of a module). So better to add a `NewTlsRESTAPI` or something like that.

A plus is that then you can make sure the user is not giving you an empty certificate path when he really means to use certificates.

Comment (dgrisham) [inline api/restapi/restapi.go]: Right, you told me that before too (about not changing the API). That's one of the big differences between writing code for production vs. personal use/school haha.

File: pnet_test.go (hunk 1 above)
Comment (dgrisham) [inline pnet_test.go]: Yeah, this was the result of me running `go fmt` (which I'd forgotten to do before). Easy enough to change, I'll make another PR with this fix.

**Comment by dgrisham** [comment]
@hsanjuan Just added a few tests for this feature, most of the functionality should be tested via the `NewRESTAPI()` tests (which shares a common helper function with `NewTLSRESTAPI()` that does most of the work). The tests I wrote primarily test the `tls.LoadX509KeyPair()` function, which obviously isn't really testing our code. There's a bit more happening in the `ipfs-cluster-service/main.go:run()` function that might be good to test in some way, not quite sure how far to go here.

File: api/restapi/restapi.go [hunk 3]
```diff
@@ -77,7 +92,30 @@ func NewRESTAPI(apiMAddr ma.Multiaddr) (*RESTAPI, error) {
 	if err != nil {
 		return nil, err
 	}
+	return newRESTAPI(apiMAddr, l)
+}
+
+// NewTlsRESTAPI creates a new REST API component that uses TLS for security
+// (authentication, encryption). It receives the multiaddress on which the API
+// listens, as well as paths to certificate and private key files
+func NewTLSRESTAPI(apiMAddr ma.Multiaddr, certFile, keyFile string) (*RESTAPI, error) {
```
Comment (hsanjuan) [inline api/restapi/restapi.go]: Let this function receive a tls.Config directly: cluster integrator's responsability to set it up.

For ipfs-cluster-service, we are the integrator, so you have to set it up there, :)

File: api/restapi/restapi.go [hunk 4]
```diff
@@ -36,6 +38,19 @@ var (
 	RESTAPIServerIdleTimeout = 120 * time.Second
 )
 
+// copied from https://github.com/denji/golang-tls
+var tlsCfg = &tls.Config{
```
Comment (hsanjuan) [inline api/restapi/restapi.go]: Move to ipfs-cluster-service/main.go (see comment below).

File: config.go [hunk 5]
```diff
@@ -66,6 +66,21 @@ type Config struct {
 	// Listen parameters for the the Cluster HTTP API component.
 	APIAddr ma.Multiaddr
 
+	// If APIUseTLS is `true`, then Cluster will use TLS (configured via the
```
Comment (hsanjuan) [inline config.go]: Redundant. Let ipfs-cluster-service use TLS if either TLSCertFile or TLSKeyFile != "".

File: ipfs-cluster-service/main.go [hunk 6]
```diff
@@ -263,7 +263,15 @@ func run(c *cli.Context) error {
 		cfg.AllocationStrategy = a
 	}
 
-	api, err := restapi.NewRESTAPI(cfg.APIAddr)
+	var api *restapi.RESTAPI
+	if cfg.APIUseTLS {
```
Comment (hsanjuan) [inline ipfs-cluster-service/main.go]: Ok, following from above, here you can check if the CertFile or KeyFile are defined. Attempt to create the TLS Config and then pass it to NewTLSRESTAPI.

**Comment by dgrisham** [comment]
@hsanjuan This last push should address all of the comments you made, assigned from the self-signed/sharness points. I tested with `ipfs-cluster-ctl --https`and everything that I could test behaved as expected (basically, either client/server forcing TLS while other wasn't would result in a failed connection). But you're right, I couldn't test both using TLS since I'm using a self-signed key.

I looked a bit and found this: https://golang.org/src/crypto/tls/common.go#L441. Might be the way to go, but we can also talk about this/look around for other approaches.

File: ipfs-cluster-service/main.go (hunk 6 above)
Comment (dgrisham) [inline ipfs-cluster-service/main.go]: Awesome, this is much cleaner.

File: ipfs-cluster-ctl/main.go [hunk 7]
```diff
@@ -85,6 +87,10 @@ func main() {
 			Name:  "https, s",
 			Usage: "use https to connect to the API",
 		},
+		cli.BoolFlag{
+			Name:  "skip-verify-cert",
```
Comment (hsanjuan) [inline ipfs-cluster-ctl/main.go]: let's use `--no-check-certificate`, that's what `wget` calls it.

Comment (dgrisham) [inline ipfs-cluster-ctl/main.go]: Fixed, and yeah we can do sharness in a different PR.
